[ZEPPELIN-6294] Made "Clone paragraph" button work properly in New UI

voidmatcha opened the pull request:
### What is this PR for?
**Description:**
Clicking the paragraph control’s Clone paragraph button that trigger COPY_PARAGRAPH. Upon receiving PARAGRAPH_ADDED, the Classic UI includes data.paragraph.text as expected, but the New UI receives an empty data.paragraph.text. Additionally, after cloning, the cursor is not positioned in the cloned paragraph in the New UI.

**Expected:**
- Cloned paragraph contains the original text.
- Cursor focuses/positions inside the cloned paragraph.

**Actual (New UI):**
- data.paragraph.text is empty on PARAGRAPH_ADDED.
- Cursor not focused / not moved to the cloned paragraph.

**[Appropriate action - Classic UI]**

https://github.com/user-attachments/assets/dd96ceaa-b0f1-4b1e-ad11-46560a0e2b79

**[AS-IS]**

https://github.com/user-attachments/assets/39530005-b169-4301-9416-ed14cb6888fc

**[TO-BE]**

https://github.com/user-attachments/assets/81302edb-0d52-49e9-a554-e2509a7bca18


### What type of PR is it?
Bug Fix

### Todos

### What is the Jira issue?
* [[ZEPPELIN-6294](https://issues.apache.org/jira/browse/ZEPPELIN-6294)]

### How should this be tested?

### Screenshots (if appropriate)

### Questions:
* Does the license files need to update? N
* Is there breaking changes for older versions? N
* Does this needs documentation? N

voidmatcha committed:
[ZEPPELIN-6294] fix: "Clone paragraph" button work properly in New UI

tbonelee reviewed (commented):
Thanks for taking a look! From my side, data.paragraph.text appears on PARAGRAPH_ADDED in both UIs, but I might be mistaken. If there’s a case where it’s empty, I’d love the exact repro steps.
One possibility is that the New UI is skipping a valid PARAGRAPH as if it were short-circuited when another request bumps lastMsgIdSeq before the final reply—`PARAGRAPH` message (see message.ts, L187). I noticed `EDITOR_SETTING` messages being sent from the client and suspect they might be the trigger. If the only intended short-circuit is RUN_PARAGRAPH → synthetic PARAGRAPH_STATUS, would it make sense to remove the second filter() at L187 rather than adding a special-case branch at the `NotebookServer`? Happy to adjust based on your feedback.

voidmatcha committed:
[ZEPPELIN-6294] fix: revert server code, adjust timing of EDITOR_SETTING

dididy commented:
I really appreciate about your review. 👍 👍

As you mentioned, the issue was caused by EDITOR_SETTING being triggered before PARAGRAPH, which led to short-circuit processing. However, modifying the condition in [message.ts, L187](https://github.com/dididy/zeppelin/blob/6ffd605bd71e3857f5680d2de26ad71469f6b198/zeppelin-web-angular/projects/zeppelin-sdk/src/message.ts#L186-L198) didn’t seem optimal from a performance standpoint.

So I adjusted the timing in the relevant section of the [app/pages/workspace/notebook/paragraph/code-editor/code-editor.component.ts](https://github.com/apache/zeppelin/pull/5044/commits/1bbaeaf1fbba60531e774a199e9c445b6242c99d#diff-b9e215ed5d0d59a7732256f1da6be620f4cdcc62019f8807a9046e60017551ae) instead. Because the [`getEditorSetting`](https://github.com/dididy/zeppelin/blob/6ffd605bd71e3857f5680d2de26ad71469f6b198/zeppelin-web-angular/src/app/pages/workspace/notebook/paragraph/code-editor/code-editor.component.ts#L196) method that triggers [EDITOR_SETTING is only executed when the parameter of `setParagraphMode` is `true`](https://github.com/dididy/zeppelin/blob/6ffd605bd71e3857f5680d2de26ad71469f6b198/zeppelin-web-angular/src/app/pages/workspace/notebook/paragraph/code-editor/code-editor.component.ts#L192-L198), and the only place it is used in this form is inside [`onDidChangeModelContent`](https://github.com/dididy/zeppelin/blob/6ffd605bd71e3857f5680d2de26ad71469f6b198/zeppelin-web-angular/src/app/pages/workspace/notebook/paragraph/code-editor/code-editor.component.ts#L86-L99). Therefore, [I modified it so that when `setParagraphMode(true)` is called in that area](https://github.com/apache/zeppelin/pull/5044/commits/1bbaeaf1fbba60531e774a199e9c445b6242c99d#diff-b9e215ed5d0d59a7732256f1da6be620f4cdcc62019f8807a9046e60017551aeR95-R98), the timing is adjusted using setTimeout.

After reverted NotebookServer code, everything worked as intended. Thanks for pointing out a better way to improve it.

voidmatcha committed:
[fix/ZEPPELIN-6294] chore: changed setTimeout position

voidmatcha committed:
[ZEPPELIN-6294] chore: revert function position

voidmatcha committed:
[ZEPPELIN-6294] chore: revert setTimeout delay time

tbonelee reviewed (changes requested):
Nice fix! Since the `setTimeout(..., 200)` is a heuristic to dodge a race, could we add a short comment explaining (1) why it's needed and (2) a TODO o how to make this deterministic or order-independent?

voidmatcha committed:
[ZEPPELIN-6294] fix: Ensure the execution order of EDITOR_SETTING

tbonelee commented on zeppelin-web-angular/src/app/core/paragraph-base/paragraph-base.ts:
```diff
@@ -54,6 +55,7 @@ export abstract class ParagraphBase extends MessageListenersManager {
 
   // Initialized by `ViewChildren` in the class which extends ParagraphBase
   notebookParagraphResultComponents!: QueryList<NotebookParagraphResultComponent>;
+  notebookParagraphCodeEditorComponent!: NotebookParagraphCodeEditorComponent;
```
I think this non-null assertion should be removed and changed to optional. I found it to be `undefined` when `paragraph.config.editorHide === true` in the `%md` editor case. Maybe you could check some errors after running `%md` interpreter
I mistakenly added non-null assertion for `notebookParagraphCodeEditorComponent` in `NotebookParagraphComponent` in a previous commit, but that was incorrect. Sorry for the confusion.

tbonelee commented on zeppelin-web-angular/src/app/pages/workspace/notebook/notebook.component.ts:
```diff
@@ -109,6 +109,10 @@ export class NotebookComponent extends MessageListenersManager implements OnInit
   @MessageListener(OP.INTERPRETER_BINDINGS)
   loadInterpreterBindings(data: MessageReceiveDataTypeMap[OP.INTERPRETER_BINDINGS]) {
+    this.listOfNotebookParagraphComponent.forEach(item => {
+      item.notebookParagraphCodeEditorComponent.editorSettingTriggerAllowed = true;
```
We should handle some case where `notebookParagraphCodeEditorComponent` is undefined.
Also, should we trigger editor settings for all paragraphs here? Would calling `this.setParagraphMode(true)` in `editor.onDidChangeModelContent()` be sufficient to get editor settings only to the paragraph whose interpreter has changed?

dididy replied:
In testing, the interpreter setting works correctly when `this.setParagraphMode(true)` is executed inside `editor.onDidChangeModelContent()`.

Since this runs **before** the code that executes when receiving `INTERPRETER_BINDINGS` above (`editor.onDidChangeModelContent()`), the code that runs after `editorSettingTriggerAllowed = true` was modified to directly call `getEditorSetting()` instead of using `setParagraphMode(true)` when receive `PARAGRAPH` or `INTERPRETER_BINDINGS`. Because if the interpreter is already set inside `this.setParagraphMode(true)`, `getEditorSetting()` will not be executed due to the condition, and since `setParagraphMode(true)` has already run, there is no need to execute it again.

tbonelee commented on zeppelin-web-angular/src/app/core/paragraph-base/paragraph-base.ts:
```diff
@@ -142,6 +144,8 @@ export abstract class ParagraphBase extends MessageListenersManager {
         }
         this.cdr.markForCheck();
       });
+      this.notebookParagraphCodeEditorComponent.editorSettingTriggerAllowed = true;
```
We should handle the case where `this.notebookParagraphCodeEditorComponent` is undefined

tbonelee commented on zeppelin-web-angular/src/app/pages/workspace/notebook/paragraph/code-editor/code-editor.component.ts:
```diff
@@ -61,6 +61,7 @@ export class NotebookParagraphCodeEditorComponent implements OnChanges, OnDestro
   private monacoDisposables: IDisposable[] = [];
   height = 18;
   interpreterName?: string;
+  editorSettingTriggerAllowed: boolean = false;
```
Could we perhaps add a comment explaining  why this property is introduced here?

tbonelee reviewed (changes requested):
Thanks for the thorough update — I’ve left some questions and requested changes.

voidmatcha committed:
[ZEPPELIN-6294] fix: set notebookParagraphCodeEditorComponent as optional

voidmatcha committed:
[ZEPPELIN-6294] fix: handle so that it also works in the INSERT_PARAGRAPH case

voidmatcha committed:
[ZEPPELIN-6294] chore: add a comment for editorSettingTriggerAllowed

dididy commented:
I rebased due to conflicts caused by merged PRs.

tbonelee commented on zeppelin-web-angular/src/app/pages/workspace/notebook/notebook.component.ts:
```diff
@@ -142,15 +155,19 @@ export class NotebookComponent extends MessageListenersManager implements OnInit
+    definedNote.paragraphs.splice(data.index, 0, data.paragraph);
+    const paragraphIndex = definedNote.paragraphs.findIndex(p => p.id === data.paragraph.id);
+
+    definedNote.paragraphs[paragraphIndex].focus = true;
+    const addedParagraph = this.listOfNotebookParagraphComponent.find((_, index) => index === paragraphIndex)
+      ?.notebookParagraphCodeEditorComponent;
```
I think something like `this.cdr.detectChanges()` should be called, so that the paragraph component is added for the new paragraph in `definedNote.paragraph`.

dididy replied:
5aab81a

Upon review, components using `ChangeDetectionStrategy.OnPush`—as you mentioned—do not detect changes in the same reference array unless the reference itself is replaced. Since I are currently modifying the array with `splice` (which does change the reference in place), Angular does not automatically recognize the change. Therefore, calling `detectChanges` or `markForCheck` is necessary.

The reason why Clone Paragraph or Add Paragraph (Insert Paragraph) in the current PR appeared to work correctly is due to Angular’s behavior of updating the view during events triggered within the component. Consequently, when executed via a button click, the focus handling works as expected, but when triggered via a shortcut key, the focus does not update properly.

To address this, after receiving `PARAGRAPH_ADDED` and performing a `splice` to update the array, we now explicitly call `detectChanges` to force change detection, which has been confirmed to work correctly.

1cdff50

Even after handling it this way, the focus event did not work correctly when triggered via shortcut keys. To address this, in `paragraph.component.ts`, we added a blur event on the current paragraph when executing `cloneParagraph` and `insertParagraph`.

---

I rebased again due to conflicts caused by merging the latest PR.

voidmatcha committed:
[ZEPPELIN-6294] fix: add detectChanges for addParagraph triggered by shortcut

voidmatcha committed:
[ZEPPELIN-6294] fix: execute blur event when call addParagraph relate functions

cloneParagraph, insterParagraph

voidmatcha committed:
[ZEPPELIN-6294] fix: change order of detectChanges because cloneParagraph is not work properly

The clone paragraph functionality wasn’t working correctly at the current position, so I moved it.

voidmatcha commented:
This has been improved with a better approach in #5254. For the cursor-related issue, I plan to create a separate issue and address it separately. - ZEPPELIN-6429
